lib/iris/_concatenate.py

```diff
@@ -281,14 +288,135 @@ class _CoordExtent(namedtuple("CoordExtent", ["points", "bounds"])):
     __slots__ = ()
 
 
+def _arrayhash(x: np.ndarray) -> np.ndarray:
+    """Compute a hash from a numpy array."""
+    return np.frombuffer(xxh3_64_digest(x.tobytes()), dtype=np.int64)
+
+
+def _chunk(x: np.ndarray, axis: int, keepdims: bool) -> np.ndarray:
+    """Compute a hash from a numpy array and keep the array dimensions."""
+    return _arrayhash(x).reshape((1,) * x.ndim)
+
+
+def _aggregate(x: np.ndarray, axis: int, keepdims: bool) -> np.int64:
+    """Compute a hash from a numpy array."""
+    return _arrayhash(x)[0]
+
+
+def _hash_array(a: da.Array | np.ndarray) -> np.int64:
+    """Calculate a hash representation of the provided array.
+
+    Calculates a 64-bit non-cryptographic hash of the provided array, using
+    the fast ``xxhash`` hashing algorithm.
+
+    Note that the computed hash depends on how the array is chunked.
+
+    Parameters
+    ----------
+    a :
+        The array that requires to have its hexdigest calculated.
+
+    Returns
+    -------
+    np.int64
+        The array's hash.
+
+    """
+    if isinstance(a, da.Array):
+        return da.reduction(
+            a,
+            chunk=_chunk,
+            combine=_chunk,
+            aggregate=_aggregate,
+            keepdims=False,
+            meta=np.empty(tuple(), dtype=np.int64),
+            dtype=np.int64,
+        )
+    return _aggregate(a, 0, False)
+
+
+class _ArrayHash(namedtuple("ArrayHash", ["value", "chunks"])):
+    """Container for a hash value and the chunks used when computing it.
+
+    Parameters
+    ----------
+    value : :class:`np.int64`
+        The hash value.
+    chunks : tuple
+        The chunks the array had when the hash was computed.
+    """
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            raise TypeError(f"Unable to compare {repr(self)} to {repr(other)}")
+        if self.chunks != other.chunks:
+            raise ValueError(
+                "Unable to compare arrays with different chunks: "
+                f"{self.chunks} != {other.chunks}"
+            )
+        return self.value == other.value
+
+
+def array_id(array: np.ndarray | da.Array) -> str:
+    """Get a deterministic token representing `array`."""
```

Comment:
As an alternative to using the tokens to look up the hashes in the dict, I could also attach the hashes as a new attribute to `_CoordAndDims` and `_DerivedCoordAndDims` instances. That would require some special care to create the derived coordinates only once (i.e. create the `_CubeSignature`s in the `concatenate` function instead of in `_ProtoCube.register`), but maybe that is a good idea anyway because it would improve performance. The downside of this alternative approach would be that it requires more changes to the code, but on the plus side, it may be conceptually simpler to understand and run faster, especially if there are derived coordinates.

Reply:
I completely removed this and replaced it by much simpler keys in f481d55.